### DISCUSSION TITLE
model-viewer not working inside the WKWebView on iOS

### DISCUSSION THREAD

**Discussion by Mikilll94** [google/model-viewer#2193] [Q&A]
### Description

Is `model-viewer` working inside the WKWebView? It works fine in Safari browser or inside SFSafariViewController, but I cannot make it work inside WKWebView?

### Browser Affected

- [ ] Chrome
- [ ] Edge
- [ ] Firefox
- [ ] Helios
- [ ] IE
- [ ] Safari
- [x] WKWebView

### OS

- [ ] Android
- [x] iOS
- [ ] Linux
- [ ] MacOS
- [ ] Windows

**Comment by cdata** [comment]
@Mikilll94 thanks for reporting. Unfortunately, we do not have testing visibility into `WKWebView` at this time.

In general, I am comfortable stating that `<model-viewer>` is broadly compatible with modern, standards-abiding web browsers. Webviews on Android and iOS are not true web browsers:

 - They don't get timely updates the same way that modern browsers do
 - They usually cannot be debugged using remote debugging without writing a lot of glue code
 - They can only be tested in CI if we create/maintain a native app ourselves

For the above reasons, we cannot offer explicit support for `<model-viewer>` in Android/iOS webviews. However, I will try to help to the extent that I can.

For troubleshooting purposes:

 - Would you be able to confirm the version information of the WKWebView you are using? (if you can get the value of `navigator.userAgent`, that would tell us what we need to know)
 - Can you confirm that you are including ALL of the recommended [polyfills](https://github.com/google/model-viewer/blob/master/packages/model-viewer/POLYFILLS.md) in the page before `<model-viewer>` is loaded?

**Comment by Mikilll94** [comment]
@cdata 

Here is the user agent of the WKWebView which I am using:

`
Mozilla/5.0 (iPhone; CPU iPhone OS 13_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148
`

**Comment by cdata** [comment]
@Mikilll94 thanks. Do you have any kind of logging set up for your app? Can you share any JavaScript exceptions that may be related?

**Comment by uprealdigital** [comment]
Hi! Any news on that?

The company 8th wall states that WebAR now works with your favorite apps like Instagram which uses WKWebView with the new iOS 14.3+.
https://www.8thwall.com/blog/post/41168830873/8th-wall-now-works-across-all-major-ios-apps-including-instagram-snapchat-and-more

How is it possible that it works for 8th wall but not with the model viewer?

**Reply by elalish** [reply]
model-viewer uses WebXR (which  doesn't exist on iOS yet) or Quick Look, which is a separate app. 8th wall works in the browser because they  do AR in JS without  ARKit at all, which is why it is slower and has a different feature set.

**Reply by uprealdigital** [reply]
@elalish Thank you so much for the valuable information and for answering my question so quickly!
